cli/internal/publish/plugins.go

```diff
@@ -26,10 +43,115 @@ type PackageJSONV1 struct {
 }
 
 type TargetBuild struct {
-	OS       string `json:"os"`
-	Arch     string `json:"arch"`
-	Path     string `json:"path"`
-	Checksum string `json:"checksum"`
+	OS             string `json:"os"`
+	Arch           string `json:"arch"`
+	Path           string `json:"path"`
+	Checksum       string `json:"checksum"`
+	DockerImageTag string `json:"docker_image_tag"`
+}
+
+type LoadResponse struct {
+	Stream string `json:"stream"`
+}
+
+type dockerProgressReader struct {
+	decoder         *json.Decoder
+	bar             *progressbar.ProgressBar
+	layerPushedByID map[string]int64
+	totalBytes      int64
+}
+
+type dockerProgressInfo struct {
+	Status       string `json:"status"`
+	Progress     string `json:"progress"`
+	ProgressData struct {
+		Current int64 `json:"current"`
+		Total   int64 `json:"total"`
+	} `json:"progressDetail"`
+	LayerID     string `json:"id"`
+	ErrorDetail struct {
+		Message string `json:"message"`
+	} `json:"errorDetail"`
+}
+
+type transportWithRegistryAuth struct {
+	http.RoundTripper
+	baseTransport *http.Transport
+	registryAuth  string
+}
+
+func newTransportWithRegistryAuth(insecureSkipVerify bool, registryAuth string) *transportWithRegistryAuth {
+	baseTransport := http.DefaultTransport.(*http.Transport).Clone()
+	baseTransport.TLSClientConfig.InsecureSkipVerify = insecureSkipVerify
+	return &transportWithRegistryAuth{
+		baseTransport: baseTransport,
+		registryAuth:  registryAuth,
+	}
+}
+
+func (t *transportWithRegistryAuth) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	req.Header.Set("Authorization", "Bearer "+t.registryAuth)
```

Comment:
Which endpoints is this for? The Docker auth expects `Basic` not `Bearer`.

Reply:
The code does basic auth with the auth endpoint to get an access token which then it uses as a `Bearer` token with the registry. This is described in https://distribution.github.io/distribution/spec/auth/token/#how-to-authenticate.

The TLDR: you start with the request you'd like to do, then if auth is required you get a challenge back to do the basic auth. With the basic auth you get a bearer token you can perform actions wiht.

1. Challenge request https://github.com/erezrokah/cloudquery/blob/6086d56a7a4f9d07bfc2e8ec2684776cad1537cb/cli/internal/publish/plugins.go#L389
2. Do basic auth https://github.com/erezrokah/cloudquery/blob/6086d56a7a4f9d07bfc2e8ec2684776cad1537cb/cli/internal/publish/plugins.go#L417
3. Get token from auth response https://github.com/erezrokah/cloudquery/blob/6086d56a7a4f9d07bfc2e8ec2684776cad1537cb/cli/internal/publish/plugins.go#L435
4. Use token to contact registry https://github.com/erezrokah/cloudquery/blob/6086d56a7a4f9d07bfc2e8ec2684776cad1537cb/cli/internal/publish/plugins.go#L468